File: CaisseEnregistreuse/CaisseEnregistreuse/InMemoryCatalog.cs
using System.Collections.Immutable;
using System.Linq;

namespace CashRegister
{
    public class InMemoryCatalog : IPriceQuery
    {
        private readonly ImmutableArray<ItemReference> _itemReferences;
        
        public InMemoryCatalog(params ItemReference[] references)
        {
            _itemReferences = references.ToImmutableArray();
        }

        public Result FindPrice(string itemCode)
        {
            var item = _itemReferences
                .Where(itemReference => itemReference.MatchByItemCode(itemCode))
                .Select(p => Result.Found(p.Price))
                .SingleOrDefault();
            return item ?? Result.NotFound(itemCode);
        }
    }
}

File: src/main/java/fr/lacombe/InMemoryCatalog.java
package fr.lacombe;

import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;

class InMemoryCatalog implements PriceQuery {
    private final Map<String, Price> unitPriceByItemCode;

    InMemoryCatalog(ItemReference... itemReferences) {
        this.unitPriceByItemCode = Stream.of(itemReferences)
                .collect(Collectors.toMap(
                        ItemReference::getItemCode,
                        ItemReference::getItemPrice
                ));
    }

    @Override
    public Result findPrice(String itemCode) {
        Price price = unitPriceByItemCode.get(itemCode);
        return price != null ? Result.found(price) : Result.notFound(itemCode);
    }

}
